File: 20/15.clj
(defn fmap [f m]
  (zipmap (keys m) (map f (vals m))))

(defn round [{i :i seen :seen prev :prev}]
  {:i (inc i)
   :seen (assoc seen prev (dec i))
   :prev (if (contains? seen prev) (- i 1 (get seen prev)) 0)})

(def nums (map read-string (clojure.string/split (slurp "15.txt") #",")))

(doseq [k [2020 30000000]]
  (time (as-> {:i (count nums)
               :seen (into {} (map vector (butlast nums) (range)))
               :prev (last nums)} $
          (iterate round $)
          (nth $ (- k (count nums)))
          (:prev $)
          (println $))))

File: 20/15.py
ns = list(map(int, open("15.txt").read().split(",")))
for k in (2020, 30000000):
    seen = {n: i for i, n in enumerate(ns[:-1])}
    prv = ns[-1]
    for i in range(len(ns), k):
        nxt = i - 1 - seen[prv] if prv in seen else 0
        seen[prv] = i - 1
        prv = nxt
    print(prv)
